### PR TITLE
make conditions_for_solution() support untrusted input

### PR DESCRIPTION
**Pull request by arvidn**
This patch fixes `parse_sexp_to_condition()` to convert to python as it parses, rather than converting the entire tree to python types up-front. This means we can fail parsing and cut argument lists short early.

Without this patch, it's not safe to use these functions on untrusted input, we don't use these in the full node for instance. However, hardening this function against untrusted input protects against future mistakes where it may be assumed to handle it.

### CODE REVIEW

**Review by cmmarslender — DISMISSED** [review]
Makes sense to me, and looks good in general. I don't have a super in depth knowledge of all the opcodes to say for certain that "oh yeah, 4 arg limit is correct" - but otherwise, the rest looks good to me

File: chia/util/condition_tools.py [hunk 1]
```diff
@@ -25,12 +25,28 @@ def parse_sexp_to_condition(
...
+    # hints and memos with this function
+    vars: List[bytes] = []
+    for arg in i:
+        a = arg.atom
+        if a is None:
+            break
+        vars.append(a)
+        # no condition (currently) has more than 3 arguments. Additional
+        # arguments are allowed but ignored
+        if len(vars) > 3:
```
Comment (cmmarslender) [inline chia/util/condition_tools.py:46]: Should this actually be `>= 3` if we only support 3 args? Currently, this ends up with a max of 4 I believe

**Review by richardkiss — DISMISSED** [review]

File: chia/util/condition_tools.py [hunk 2]
```diff
@@ -25,12 +25,28 @@ def parse_sexp_to_condition(
     Takes a ChiaLisp sexp and returns a ConditionWithArgs.
     If it fails, returns an Error
     """
-    as_atoms = sexp.as_atom_list()
-    if len(as_atoms) < 1:
+    i = sexp.as_iter()
```
[the remark this answers is not in the document]
Comment (arvidn) [reply chia/util/condition_tools.py]: I made this change

File: chia/util/condition_tools.py [hunk 3]
```diff
@@ -25,12 +25,28 @@ def parse_sexp_to_condition(
     Takes a ChiaLisp sexp and returns a ConditionWithArgs.
     If it fails, returns an Error
     """
-    as_atoms = sexp.as_atom_list()
-    if len(as_atoms) < 1:
+    first = sexp.pair
```
Comment (richardkiss) [inline chia/util/condition_tools.py:28]: The name `first` is a bit misleading, since it's actually a pair. Really, `first` is `sexp.pair[0]`.

**Review by richardkiss — APPROVED** [review]
I approved it anyway.

**Review by emlowe — APPROVED** [review]

**Pull request merged**